### PR TITLE
Fix registration for GatherGrad

### PR DESCRIPTION
**Pull request by kit1980**
Fix registration for GatherGrad to fix GatherGradOpTest.GatherGrad_axis0_indices2d_half.

### CODE REVIEW

**Comment by ashbhandare** [comment]
Can you also fix the schema definition at orttraining/orttraining/core/graph/training_op_defs.cc

File: orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc [hunk 1]
```diff
@@ -16,7 +16,8 @@ ONNX_OPERATOR_KERNEL_EX(
     KernelDefBuilder()
         .InputMemoryType<OrtMemTypeCPUInput>(0)
         .TypeConstraint("I", DataTypeImpl::GetTensorType<int64_t>())
-        .TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes())
+        .TypeConstraint("T", {DataTypeImpl::GetTensorType<float>(),
+                              DataTypeImpl::GetTensorType<MLFloat16>()})
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:20]: CUDA should support float16

Comment (kit1980) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:20]: > CUDA should support float16

`MLFloat16` is float16, no?
The current implementation only works for `float` and `MLFloat16`, that's why I changed the registration to these types:
https://github.com/microsoft/onnxruntime/blob/45dae88a1ad7c4878f4ba41ad54e5b0c069076e8/orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc#L77

**Review by SherlockNoMad — APPROVED** [review]

**Comment by kit1980** [comment]
> Can you also fix the schema definition at orttraining/orttraining/core/graph/training_op_defs.cc, and the cuda registration at orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc which is also registering for extra ops.

Done.

**Review by ashbhandare — APPROVED** [review]

**Pull request merged**